File: register_server.py
#!/bin/env python

import requests
import json
import time

from threading import Thread, Lock

from wyliodrin import *


DIGITAL_INPUT   = "DIGITAL_INPUT"
DIGITAL_OUTPUT  = "DIGITAL_OUTPUT"
PWM_OUTPUT      = "PWM_OUTPUT"
ANALOG_INPUT    = "ANALOG_INPUT"
GENERIC_OUTPUT  = "GENERIC_OUTPUT"
GENERIC_INPUT   = "GENERIC_INPUT"



class Sensor(object):

    def __init__(self, _id, _type, pin, sleep_time, value_range, function = None):
        self._id = _id
        self._type = _type
        self.lock = Lock()
        self.value = 0
        self.pin = pin
        self.value_range = value_range
        self.sleep_time = 0 if sleep_time < 0 else sleep_time
        self.function = function

        if self._type == DIGITAL_INPUT:
            pinMode(pin, INPUT)
        elif self._type == DIGITAL_OUTPUT:
            pinMode(pin, OUTPUT)


    def get(self, url, token):
        url = url + ('/' if url[-1] != '/' else '') + 'get'
        while True:     
            try:
                self.lock.acquire()
                
                data = {"id" : self._id, "token" : token}
                res = requests.post(url, data=data)

                if res.status_code == 200:
                    obj = json.loads(res.content)

                    if obj["error"] != 0:
                        print obj["reason"]
                    else:
                        self.value = obj["value"]
                        print "Got value: ", self.value, " for sensor on pin: ", self.pin

                        if self._type == DIGITAL_OUTPUT:
                            digitalWrite(self.pin, self.value)
                        elif self._type == PWM_OUTPUT:
                            analogWrite(self.pin, self.value)
                        elif self._type == GENERIC_OUTPUT:
                            self.function(self.value)
                else:
                    print "Error: " + res.reason

            finally:
                self.lock.release()
            time.sleep(self.sleep_time)



    def send(self, url, token):
        url = url + ('/' if url[-1] != '/' else '') + 'get'
        while True:
            try:
                self.lock.acquire()
                
                if self._type == DIGITAL_INPUT:
                    val = digitalRead(self.pin)
                    if abs(val - self.value) <= self.value_range:
                        continue
                    self.value = val
                elif self._type == ANALOG_INPUT:
                    val = analogRead(self.pin)
                    if abs(val - self.value) <= self.value_range:
                        continue
                    self.value = val
                elif self._type == GENERIC_INPUT:
                    val = self.function(0)
                    if abs(val - self.value) <= self.value_range:
                        continue
                    self.value = val

                data = {"id" : self._id, "token" : token, "value" : self.value}
                res = requests.post(url, data=data)

                if res.status_code != 200:
                    print "Error: " + res.reason
                else:
                    print "Sent value: ", self.value, " for sensor on pin: ", self.pin

            finally:
                self.lock.release()
            time.sleep(self.sleep_time)


def validate_pin(pin):
    if pin > 0:
        return True
    return False


class RegisterServer(object):

    def __init__(self, url, token):
        self.sensors = {}
        self.url = url
        self.token = token

    def registerDigitalInput(self, _id, pin, sleep_time, value_range):

        if _id in self.sensors or not validate_pin(pin):
            print "Id: ", _id, " already registerd"
            return False
        else:
            s = Sensor(_id, DIGITAL_INPUT, pin, sleep_time, value_range)
            self.sensors[_id] = {"object" : s, "autosend" : sleep_time}
    
            url = self.url + ("/" if self.url[-1] != "/" else "") + "/register"
            data = {"id" : _id, "token" : self.token, "type" : DIGITAL_INPUT}
            res = requests.post(url, data=data)

            if res.status_code != 200:
                print "Error regitstering DIGITAL_INPUT on pin: ", pin, " Reason: ", res.reason
                return False
            
            obj = json.loads(res.content)
            if obj["error"] != 0 and obj["error"] != 2:
                print "Error regitstering DIGITAL_INPUT on pin: ", pin, " Reason: ", obj["reason"]
                return False

            s.thread = Thread(target=s.send, args=(self.url, self.token))
            s.thread.start()

    def registerAnalogInput(self, _id, pin, sleep_time, value_range):

        if _id in self.sensors or not validate_pin(pin):
            print "Id: ", _id, " already registerd"
            return False
        else:
            s = Sensor(_id, ANALOG_INPUT, pin, sleep_time, value_range)
            self.sensors[_id] = {"object" : s, "autosend" : sleep_time}

            url = self.url + ("/" if self.url[-1] != "/" else "") + "/register"
            data = {"id" : _id, "token" : self.token, "type" : ANALOG_INPUT}
            res = requests.post(url, data=data)

            if res.status_code != 200:
                print "Error regitstering ANALOG_INPUT on pin: ", pin, " Reason: ", res.reason
                return False
            
            obj = json.loads(res.content)
            if obj["error"] != 0 and obj["error"] != 2:
                print "Error regitstering ANALOG_INPUT on pin: ", pin, " Reason: ", obj["reason"]
                return False

            s.thread = Thread(target=s.send, args=(self.url, self.token))
            s.thread.start()

    def registerDigitalOutput(self, _id, pin, sleep_time, value_range):

        if _id in self.sensors or not validate_pin(pin):
            print "Id: ", _id, " already registerd"
            return False
        else:
            s = Sensor(_id, DIGITAL_OUTPUT, pin, sleep_time, value_range)
            self.sensors[_id] = {"object" : s, "autosend" : sleep_time}

            url = self.url + ("/" if self.url[-1] != "/" else "") + "/register"
            data = {"id" : _id, "token" : self.token, "type" : DIGITAL_OUTPUT}
            res = requests.post(url, data=data)

            if res.status_code != 200:
                print "Error regitstering DIGITAL_OUTPUT on pin: ", pin, " Reason: ", res.reason
                return False
            
            obj = json.loads(res.content)
            if obj["error"] != 0 and obj["error"] != 2:
                print "Error regitstering DIGITAL_OUTPUT on pin: ", pin, " Reason: ", obj["reason"]
                return False

            s.thread = Thread(target=s.get, args=(self.url, self.token))
            s.thread.start()

    def registerPWMOutput(self, _id, pin, sleep_time, value_range):

        if _id in self.sensors or not validate_pin(pin):
            print "Id: ", _id, " already registerd"
            return False
        else:
            s = Sensor(_id, PWM_OUTPUT, pin, sleep_time, value_range)
            self.sensors[_id] = {"object" : s, "autosend" : sleep_time}

            url = self.url + ("/" if self.url[-1] != "/" else "") + "/register"
            data = {"id" : _id, "token" : self.token, "type" : PWM_OUTPUT}
            res = requests.post(url, data=data)

            if res.status_code != 200:
                print "Error regitstering PWM_OUTPUT on pin: ", pin, " Reason: ", res.reason
                return False
            
            obj = json.loads(res.content)
            if obj["error"] != 0 and obj["error"] != 2:
                print "Error regitstering PWM_OUTPUT on pin: ", pin, " Reason: ", obj["reason"]
                return False

            s.thread = Thread(target=s.get, args=(self.url, self.token))
            s.thread.start()


    def registerGenericOutput(self, _id, pin, sleep_time, value_range, function):

        if _id in self.sensors or not validate_pin(pin):
            print "Id: ", _id, " already registerd"
            return False
        else:
            s = Sensor(_id, GENERIC_OUTPUT, pin, sleep_time, value_range, function)
            self.sensors[_id] = {"object" : s, "autosend" : sleep_time}

            url = self.url + ("/" if self.url[-1] != "/" else "") + "/register"
            data = {"id" : _id, "token" : self.token, "type" : GENERIC_OUTPUT}
            res = requests.post(url, data=data)

            if res.status_code != 200:
                print "Error regitstering GENERIC_OUTPUT on pin: ", pin, " Reason: ", res.reason
                return False
            
            obj = json.loads(res.content)
            if obj["error"] != 0 and obj["error"] != 2:
                print "Error regitstering GENERIC_OUTPUT on pin: ", pin, " Reason: ", obj["reason"]
                return False

            s.thread = Thread(target=s.get, args=(self.url, self.token))
            s.thread.start()

    def registerGenericInput(self, _id, pin, sleep_time, value_range, function):

        if _id in self.sensors or not validate_pin(pin):
            print "Id: ", _id, " already registerd"
            return False
        else:
            s = Sensor(_id, GENERIC_INPUT, pin, sleep_time, value_range, function)
            self.sensors[_id] = {"object" : s, "autosend" : sleep_time}

            url = self.url + ("/" if self.url[-1] != "/" else "") + "/register"
            data = {"id" : _id, "token" : self.token, "type" : GENERIC_INPUT}
            res = requests.post(url, data=data)

            if res.status_code != 200:
                print "Error regitstering GENERIC_INPUT on pin: ", pin, " Reason: ", res.reason
                return False
            
            obj = json.loads(res.content)
            if obj["error"] != 0 and obj["error"] != 2:
                print "Error regitstering GENERIC_INPUT on pin: ", pin, " Reason: ", obj["reason"]
                return False

            s.thread = Thread(target=s.send, args=(self.url, self.token))
            s.thread.start()

File: Arduino/RegisterServer.cpp
/**
 * @author Andrei Stefanescu
 * Wyliodrin SRL
 * 
 */
#include "RegisterServer.h"


String sendRequest(const char* host, const char* path, const char* data) {
  EthernetClient client;

  /* connect and send request */
  String value;
  if (client.connect(host, 80)) {

    /* send headers */
    String line = "POST ";
    line += path;
    line += " HTTP/1.1";
    String h = "Host: ";
    h += host;
    Serial.println("Connected");
    String length = "Content-Length: ";
    length += strlen(data);

    client.println(line);
    client.println("Content-Type: application/x-www-form-urlencoded");
    client.println(length);
    client.println("Connection: close");
    client.println(h);
    client.println();
    client.println(data);
    client.println();

    char old = '\0', current = '\0';
    bool mode = false;

    /* read result */
    while (client.connected()) {
      if (client.available()) {
        if (current != '\r') old = current;
        current = client.read();
        if (mode) value += current;
        
        if (current == old && current == '\n') mode = true;
      }
    }

    /* also deallocates data */
    client.stop();
  } else {
    Serial.println("Failed connecting");
  }

  /* for debugging */
  Serial.println("Got: " + value);

  return value;
}

Sensor::Sensor() {
    this->id = String("invalid");
    this->pin = -1;
    this->value = 0;
    this->type  = UNDEFINED;
    this->range = 0;
}
  
Sensor::Sensor(const String& id, sensor_type type, int pin, int range) {
    this->id = id;
    this->pin = pin;
    this->value = 0;
    this->type = type;
    this->range = range;
}

Sensor::Sensor(const Sensor& s) {
    this->id = s.id;
    this->pin = s.pin;
    this->value = s.value;
    this->type = s.type;
    this->range = s.range;
}


Sensor& Sensor::operator=(const Sensor& s) {
    this->id = s.id;
    this->pin = s.pin;
    this->value = s.value;
    this->type = s.type;
    this->range = s.range;

    return *(this);
}


Register::RegisterServer() {
    nrSensors = 0;
#ifdef NETWORKING_STATUS
    pinMode(13, OUTPUT);
#endif
}


int RegisterServer::begin(String serverAddress, byte mac[4], String token) {

    #if defined(ARDUINO_YUN_)
      Bridge.begin();  /* for internet */
      Console.begin(); /* for debugging/logging */
    #elif defined(ARDUINO_UNO_)
      /* begin ethernet */
      delay(1000);
      Serial.begin(9600);

      if (!Ethernet.begin(mac)) {
        Serial.println("Error configuring ethernet");
      }
      Serial.println("Configured");
    #endif


    this->token = token;
    this->url = serverAddress;
    this->nrSensors = 0;

    return 0;
}

bool RegisterServer::registerGenericInput(String id,
                                          int pin,
                                          int range,
                                          float (*func)(float)) {
    if (nrSensors == MAX_SIZE) {
      return false;
}
    
    /* create sensor and add it */
    Sensor s(id, GENERIC_INPUT, pin, range);
    sensors[nrSensors] = s;
    functions[nrSensors] = func;
    nrSensors += 1;
   
    String newUrl = url + "/register";
    String data = "id=" + id + "&token=" + token + "&type=GENERIC_INPUT";

    /* send request to server */
    #if defined(ARDUINO_YUN_)    
      HttpClient client;
      client.post(newUrl, data);
    #elif defined(ARDUINO_UNO_)
      sendRequest(url.c_str(), "/register", data.c_str());
    #endif
 
    return true;
}


bool RegisterServer::registerGenericOutput(String id,
                                           int pin,
                                           int range,
                                           float (*func)(float)) {
    if (nrSensors == MAX_SIZE) {
      return false;
    }
    
    /* create sensor and add it */
    Sensor s(id, GENERIC_OUTPUT, pin, range);
    sensors[nrSensors] = s;
    functions[nrSensors] = func;
    nrSensors += 1;
   
    String newUrl = url + "/register";
    String data = "id=" + id + "&token=" + token + "&type=GENERIC_OUTPUT";
    
    /* send request to server */
    #if defined(ARDUINO_YUN_)
      HttpClient client;
      client.post(newUrl, data);
    #elif defined(ARDUINO_UNO_)
      sendRequest(url.c_str(), "/register", data.c_str());
    #endif
    func(0);

    return true;
}
  

bool RegisterServer::registerDigitalInput(String id, int pin) {
    if (nrSensors == MAX_SIZE) {
      return false;
    }

    /* set the board for it */
    pinMode(pin, INPUT);
   
    /* create it */ 
    Sensor s(id, DIGITAL_INPUT, pin, 0);
    sensors[nrSensors] = s;
    nrSensors += 1;
   
    String newUrl = url + "/register";
    String data = "id=" + id + "&token=" + token + "&type=DIGITAL_INPUT";
    
    /* send request to server */
    #if defined(ARDUINO_YUN_)
      HttpClient client;
      client.post(newUrl, data);
    #elif defined(ARDUINO_UNO_)
      sendRequest(url.c_str(), "/register", data.c_str());
    #endif

    return true;
}

bool registerAnalogInput(String id, int pin, int range) {
    if (nrSensors == MAX_SIZE) {
      return false;
    }
    
    /* create it and add it */
    Sensor s(id, ANALOG_INPUT, pin, range);
    sensors[nrSensors] = s;
    nrSensors += 1;
   
    String newUrl = url + "/register";
    String data = "id=" + id + "&token=" + token + "&type=ANALOG_INPUT";

    /* send request to server */
    #if defined(ARDUINO_YUN_)
      HttpClient client;
      client.post(newUrl, data);
      Console.println("Registering aI");
    #elif defined(ARDUINO_UNO_)
      sendRequest(url.c_str(), "/register", data.c_str());
    #endif

    return true;
}


bool RegisterServer::registerDigitalOutput(String id, int pin) {
    if (nrSensors == MAX_SIZE) {
      return false;
    }
    pinMode(pin, OUTPUT);
    
    /* create and add it */
    Sensor s(id, DIGITAL_OUTPUT, pin, 0);
    sensors[nrSensors] = s;
    nrSensors += 1;
   
    String newUrl = url + "/register";
    String data = "id=" + id + "&token=" + token + "&type=DIGITAL_OUTPUT";
    
    /* send request to server */
    #if defined(ARDUINO_YUN_)
      HttpClient client;
      client.post(newUrl, data);
      Console.println("Regist do");
    #elif defined(ARDUINO_UNO_)
      sendRequest(url.c_str(), "/register", data.c_str());
    #endif
    digitalWrite(pin, 0);

    return true;
}


bool RegisterServer::registerPWMOutput(String id, int pin) {
    if (nrSensors == MAX_SIZE) {
      return false;
    }

    /* create sensor and add it */
    Sensor s(id, PWM_OUTPUT, pin, 0);
    sensors[nrSensors] = s;
    nrSensors += 1;
   
    String newUrl = url + "/register";
    String data = "id=" + id + "&token=" + token + "&type=PWM_OUTPUT";
    
    /* send request to server */
    #if defined(ARDUINO_YUN_)
      HttpClient client;
      client.post(newUrl, data);
    #elif defined(ARDUINO_UNO_)
      sendRequest(url.c_str(), "/register", data.c_str());
    #endif
    analogWrite(pin, 0);
    

    return true;
}


void RegisterServer::loop() {

    /* communicating with the server notification */
#ifdef NETWORKING_STATUS
    digitalWrite(NETWORKING_PIN, HIGH);
#endif
    /* iterate over sensors and update them */
    for (int i = 0; i < nrSensors; ++ i) {
      if (sensors[i].type == DIGITAL_INPUT ||
          sensors[i].type == ANALOG_INPUT  || 
          sensors[i].type == GENERIC_INPUT) {
        
        int val;
        
        /* retrieve value */
        if(sensors[i].type == DIGITAL_INPUT) {
          val = digitalRead(sensors[i].pin);
        } else if (sensors[i].type == ANALOG_INPUT) {
          val = analogRead(sensors[i].pin);
        } else {
          val = functions[i](0);
        }
        
        if (abs(sensors[i].value - val)  <= sensors[i].range) {
          /* new value is within range */ 
          continue;
        }
        
        /* prepare request for server */
        sensors[i].value = val;
        String data = "id=" +
                      sensors[i].id +
                      "&token=" +
                      token +
                      "&value=" +
                      sensors[i].value;

        /* send request */
        #if defined(ARDUINO_YUN_)
          HttpClient c;
          String newUrl = url + "/send";

          c.post(newUrl, data);
        #elif defined(ARDUINO_UNO_)
          sendRequest(url.c_str(), "/send", data.c_str());
        #endif

      } else if (sensors[i].type == DIGITAL_OUTPUT || 
                 sensors[i].type == PWM_OUTPUT     ||
                 sensors[i].type == GENERIC_OUTPUT) {

        /* prepare retrieval request */
        String data = "id=" + sensors[i].id + "&token=" + token + "&plain=1";
        String value = "";

        /* send request */
        #if defined(ARDUINO_YUN_)
          HttpClient c;
          
          String newUrl = url + "/get";   
          c.post(newUrl, data);

          /* read response */
          while (c.available()) {
            char w = c.read();          
            if (!isdigit(w) && w != '.') break;
            value += w;
          }

          sensors[i].value = value.toFloat();

        #elif defined(ARDUINO_UNO_)
          /* send and retrieve response */
          value = sendRequest(url.c_str(), "/get", data.c_str());
          sensors[i].value = value.toFloat();
        #endif

        /* set the sensor's value */
        if (sensors[i].type == DIGITAL_OUTPUT) {
            digitalWrite(sensors[i].pin, value.toInt());
        } else if (sensors[i].type == PWM_OUTPUT) {
            analogWrite(sensors[i].pin, value.toInt());
        } else {
            functions[i](value.toFloat());
        }
        
      }
    }

    /* no more communicating with the server */
#ifdef NETWORKING_STATUS
    digitalWrite(NETWORKING_PIN, LOW);
#endif
}


void RegisterServer::printStatus() {
    #if defined(ARDUINO_YUN_)
      Console.print("-------------------------\n");
      Console.flush();
      for (int i = 0; i < nrSensors; ++ i) {
        Console.print(sensors[i].id +
                      " has value: " +
                      String(sensors[i].value) +
                      "\n");
      }

      Console.print("-------------------------\n");
    #elif defined(ARDUINO_UNO_)
      Serial.print("-------------------------\n");
      Serial.flush();
      for (int i = 0; i < nrSensors; ++ i) {
        Serial.print(sensors[i].id + " has value: " +
                     String(sensors[i].value) + "\n");
      }

      Serial.print("-------------------------\n");
    #endif
}
